include/adm/detail/id_parser.hpp

```diff
@@ -23,19 +23,32 @@ namespace adm {
       void check_size(size_t size) {
         if (id.size() != size) {
           std::ostringstream errorString;
-          errorString << "invalid " << type << " (wrong length): " << id;
+          errorString << "invalid " << type << " (wrong length, should be "
+                      << std::to_string(size) << " characters): " << id;
           throw std::runtime_error(errorString.str());
         }
       }
 
+      void check_prefix(const char *prefix, size_t size) const {
+        assert(prefix && prefix[size] == 0);
+        check_prefix(std::string(prefix));
+      }
+
       /// check that the start of the ID matches the given prefix
-      void check_prefix(const char *prefix, size_t prefix_len) {
-        assert(prefix[prefix_len] == 0);
-        for (size_t i = 0; i < prefix_len; i++)
-          if (id[i] != prefix[i]) {
-            std::ostringstream errorString;
-            errorString << "invalid " << type << " (wrong prefix): " << id;
-            throw std::runtime_error(errorString.str());
+      void check_prefix(std::string const &prefix) const {
+        auto prefix_error = [this](std::string const &prefix) {
+          std::ostringstream errorString;
+          errorString << "invalid " << type << " (incorrect prefix, should be '"
+                      << prefix << "'): " << id;
+          throw std::runtime_error(errorString.str());
+        };
+
+        if (prefix.size() > id.size()) {
+          prefix_error(prefix);
+        }
```

Comment:
This didn't check the id length before because a call to `check_prefix` is always preceded by a call to `check_size`, so it will never fire. I'd prefer this to be an assertion (since it's a programming error), which would remove the need for the lambda.

That would be consistent with the other methods in this class too.

Reply:
Sorry, I missed the context for this before. I agree with the idea of checking the prefix first, so please ignore this.